Remove read stdin loop

ghubstan opened the pull request:
Replaced the Scanner input read loop with upgraded joptsimple
dependency. Cli now takes a single, non-option program argument, runs
it and exits.  Also removed the "stop client" command because there is
no input loop, but shutdown() is called for orderly channel shudown
before the jvm terminates.  Also changed cmd syntax from camel case
to lowercase, mimicking bitcoin-cli.

Configured logback to supress all debug & info level netty output, and
bypassed logback to print results to System.out (replaces #4091).

<!-- 
- make yourself familiar with the CONTRIBUTING.md if you have not already (https://github.com/bisq-network/bisq/blob/master/CONTRIBUTING.md)
- make sure you follow our [coding style guidelines][https://github.com/bisq-network/style/issues)
- pick a descriptive title
- provide some meaningful PR description below
- create the PR
- in case you receive a "Change request" and/or a NACK, please react within 30 days. If not, we will close your PR and it can not be up for compensation.
- After addressing the change request, __please re-request a review!__ Otherwise we might miss your PR as we tend to only look at pull requests tagged with a "review required".
-->

ghubstan committed:
Remove read stdin loop

Replaced the Scanner input read loop with upgraded joptsimple
dependency. Cli now takes a single, non-option program argument, runs
it and exits.  Also removed the "stop client" command because there is
no input loop, but shutdown() is called for orderly channel shudown
before the jvm terminates.  Also changed cmd syntax from camel case
to lowercase, mimicking bitcoin-cli.

Configured logback to supress all debug & info level netty output, and
bypassed logback to print results to System.out.

cbeams commented on cli/src/main/java/bisq/cli/app/BisqCliMain.java:
```diff
@@ -41,60 +45,28 @@
 
     private final ManagedChannel channel;
     private final CliCommand cmd;
+    private final OptionParser cmdParser;
```
```suggestion
    private final OptionParser parser;
```

cbeams commented on cli/src/main/java/bisq/cli/app/BisqCliMain.java:
```diff
@@ -41,60 +45,28 @@
 
     public static void main(String[] args) {
-        new BisqCliMain("localhost", 8888);
+        new BisqCliMain("localhost", 8888, args);
```
```suggestion
        new BisqCliMain("localhost", 9998, args);
```

8888 is a poor number as it's frequently used by other services. 9998 fits with bisq's p2p port of 9999 and follows suit with bitcoind's 8333 p2p port and 8332 rpc port. I realize you did not introduce this change and that it's somewhat out of scope for this PR, but please make it anyway.

@freimair, please note that the choice of 9998 here overlaps with the same port used in `monitor/src/main/java/bisq/monitor/metric/TorHiddenServiceStartupTime.java`. Please consider changing that to something else if you see it being a conflict.

ghubstan replied:
BisqGrpcServer now listens on port 9998.

ghubstan replied:
A reminder to @freimair, re `TorHiddenServiceStartupTime`
```
int localPort = Integer.parseInt(configuration.getProperty(LOCAL_PORT, "9998"));
int servicePort = Integer.parseInt(configuration.getProperty(SERVICE_PORT, "9999"));
```
now conflicts with the `BisqGrpcServer` listening port `9998` in this `remove-read-stdin-loop` branch (not merged).   I am sure I should not be changing server + listening ports in `TorHiddenServiceStartupTime`.

freimair replied:
I do not think so. the monitor is quite separate from everything else, plus, this is a default fallback value - it can and should be configured via the monitors properties file.

cbeams commented on cli/src/main/java/bisq/cli/app/BisqCliMain.java:
```diff
@@ -41,60 +45,28 @@
     }
 
-    private BisqCliMain(String host, int port) {
+    private BisqCliMain(String host, int port, String[] params) {
```
```suggestion
    private BisqCliMain(String host, int port, String[] args) {
```

For consistency and conventionality.

cbeams commented on cli/src/main/java/bisq/cli/app/BisqCliMain.java:
```diff
@@ -41,60 +45,28 @@
-                // First response is rather slow (300 ms) but following responses are fast (3-5 ms).
-                log.info("{}\t{}", result, cmd.responseTime.apply(startTs));
-            }
+        long startTs = currentTimeMillis();
```
Please remove this.

cbeams commented on cli/src/main/java/bisq/cli/app/BisqCliMain.java:
```diff
@@ -41,60 +45,28 @@
+
+        // First response is rather slow (300 ms) but following responses are fast (3-5 ms).
+        // log.info("{}\t{}", result, cmd.responseTime.apply(startTs));
+        out.println(result + "\t" + cmd.responseTime.apply(startTs));
```
And this as well. There is no requirement at present to measure the response time, and output from a unix utility should not include anything other than the data requested, i.e. such that it can be piped to the next utility. This sort of output breaks that unix idiom. Again, I realize you did not introduce this. I'm just asking you to remove it and start thinking about things this way, making similar decisions as you go.

cbeams commented on cli/src/main/java/bisq/cli/app/CommandParser.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    static void printUsage() {
+        err.println("Usage:  bisq-cli getbalance | getversion");
```
Please study the way help output is printed in `BisqExecutable#execute` and note the use of `BisqHelpFormatter` which we may want to reuse (we would need to move it in order to do so, don't worry about it now, just use the default help formatter implementation in the meantime).

In any case, when help is explicitly requested, it shouldn't be printed to stderr, but stdout.

Please also explicitly `System.exit` 0 or 1 accordingly. You'll see how this is done in `BisqExecutable` as well.

ghubstan replied:
Prints usage message to stdout, and uses EXIT_SUCCESS and EXIT_FAILURE arguments in System.exit(). 

What's not done:  help formatting.  I need to discuss Bisq + `joptsimple.HelpFormatter` issues with @cbeams.  Right now, the only options passed to Cli are of type `NonOptionArgumentSpec`, which did not work as expected with `joptsimple.HelpFormatter`.  It looks like Bisq Config only uses `ArgumentAcceptingOptionSpec` options, which do work with `joptsimple.HelpFormatter`.  It did not seem correct to change non-option cli arguments to argument accepting options just to make them work with the `joptsimple.HelpFormatter` interface, which is why I checked in the simple 'print usage' print statement.

cbeams replied:
Ok, that's fine then. We can come back to it later, thanks.

cbeams reviewed (changes requested):
See my comments and suggested changes. There are additional changes I'd make as well, but time is limited and this is good enough to go forward. Thanks.

ghubstan committed:
Implement requested changes in PR #4097

Change member name OptionParser cmdParser -> parser.
Change server listening port to 9998, client port to 9998.
Change constructor argument from String[] param -> args.
Print the result only, w/out exec time.
Handle help command & print that to stdout;  print help
triggered by user error to stderr.
Use explicit system SUCCESS/FAIL codes in System.exit(0 || 1).

cbeams commented:
Please have another look at how help / usage output is handled in `BisqExecutable` and co. I’m referring to hooking into JOptSimple’s built-in support here.

cbeams commented:
> Please have another look at how help / usage output is handled in `BisqExecutable` and co. I’m referring to hooking into JOptSimple’s built-in support here.

Please disregard this message. I wrote it via mobile, where I hadn't seen @ghubstan's https://github.com/bisq-network/bisq/pull/4097#discussion_r400246596.

cbeams commented:
I would generally recommend against force pushing on PR branches, especially where feedback is being folded in. It destroys history / information, and messes with merging / rebasing if someone (like myself) is coming along with review commits as well. I don't know why you did these force pushes, perhaps there was a great reason, just mentioning.

cbeams reviewed (approved):
ACK, the changes look good. Thanks @ghubstan.

ghubstan commented:
> I would generally recommend against force pushing on PR branches, especially where feedback is being folded in. It destroys history / information, and messes with merging / rebasing if someone (like myself) is coming along with review commits as well. _I don't know why you did these force pushes_, perhaps there was a great reason, just mentioning.

Sorry.  No good enough reason for squashing too many intermediate commits, just a lot of interruptions today, preventing me from correctly implementing all of your requested changes in one push before the end of your, @cbeams' day.  I thought one clean (2nd) commit message was more important.  I'll be more careful and patient next time.

cbeams commented:
All good, thanks!
